### PR TITLE
support save load optimizer master_weights

### PR DESCRIPTION
**Pull request by pangengzheng**
### PR types

Others
### PR changes

Others
### Description

card-78318
support flatten state_dict, save load optimizer master_weights and deduplicate tensor when save state_dict

### CODE REVIEW

File: python/paddle/distributed/checkpoint/utils.py [hunk 1]
```diff
@@ -61,5 +61,47 @@ def compute_local_shape_and_global_offset(
 
 
 def flatten_state_dict(state_dict):
```
Comment (zhiqiu) [inline python/paddle/distributed/checkpoint/utils.py:65]: Does it support multiple level, i.e., {'model': {'m': {'w': xxx}}} ?

File: python/paddle/optimizer/optimizer.py [hunk 2]
```diff
@@ -402,35 +402,7 @@ def set_state_dict(self, state_dict):
...
-                )
-
-                assert (
-                    model_np.dtype == load_para_np.dtype
-                ), "Parameter dtype not match, Dygraph Parameter [ {} ] need tensor with dtype {}  but load tensor with dtype {}".format(
-                    model_np.name, model_np.dtype, load_para_np.dtype
-                )
-
-                tensor.set(load_para_np, framework._current_expected_place())
+                var.set_value(state_dict[var_tmp.name])
```
Comment (zhiqiu) [inline python/paddle/optimizer/optimizer.py:409]: why change here？

File: python/paddle/distributed/checkpoint/save_state_dict.py [hunk 3]
```diff
@@ -74,6 +73,16 @@ def dedup_storage_metadata(global_storage_metadata):
     return out
 
 
+def dedup_tensor(state_dict, local_storage_metadata, dedup_storage_metadata):
```
Comment (zhiqiu) [inline python/paddle/distributed/checkpoint/save_state_dict.py]: add some comments

File: python/paddle/distributed/checkpoint/utils.py (hunk 1 above)
Comment (pangengzheng) [reply python/paddle/distributed/checkpoint/utils.py:65]: 支持的

File: python/paddle/optimizer/optimizer.py (hunk 2 above)
Comment (pangengzheng) [reply python/paddle/optimizer/optimizer.py:409]: 代码复用，set_value api的行为包含这部分删掉的代码逻辑，且支持设置distributed tensor赋值

File: python/paddle/distributed/checkpoint/save_state_dict.py (hunk 3 above)
Comment (pangengzheng) [reply python/paddle/distributed/checkpoint/save_state_dict.py]: ok，不过这个方法可以算是此文件的私有方法，不对外。comment已加

**Review by XieYunshen — APPROVED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**